pkg/reconciler/ingress/resources/virtual_service.go

```diff
@@ -133,7 +133,14 @@ func MakeVirtualServices(ing *v1alpha1.Ingress, gateways map[v1alpha1.IngressVis
 }
 
 func makeVirtualServiceSpec(ing *v1alpha1.Ingress, gateways map[v1alpha1.IngressVisibility]sets.String, hosts sets.String) *istiov1alpha3.VirtualService {
-	gw := sets.String{}.Union(gateways[v1alpha1.IngressVisibilityClusterLocal]).Union(gateways[v1alpha1.IngressVisibilityExternalIP])
+	visibility := ing.Spec.Visibility
+	if visibility == "" {
+		visibility = v1alpha1.IngressVisibilityExternalIP
+	}
+	// If the visibility is ClusterLocal, only use the local Gateways.
+	// If the visibility is ExternalIP, use both local and external Gateways.
+	gw := sets.String{}.Union(gateways[v1alpha1.IngressVisibilityClusterLocal]).Union(gateways[visibility])
```

Comment:
Can we make `gw` the union of all `spec.Http.Gateways`, i.e. appending new gateways after line #154, and assign `spec.Gateways` to `gw` before returning?

That'll ensure we only declare what we need.  Alternatively I think we could leave it empty, which may be confusing for users exploring VirtualService using `kubectl get virtualservice`.

Reply:
@tcnghia you're right, it's better this way because we only get the Gateways that are used.
I will rewrite the code to add all Gateways used in `spec.Http.Match[*].Gateways`.